File: java/p174.java
/* 
 * Solution to Project Euler problem 174
 * Copyright (c) Project Nayuki. All rights reserved.
 * 
 * https://www.nayuki.io/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p174 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p174().run());
	}
	
	
	private static final int SIZE_LIMIT = 1000000;
	private static final int TYPE_LIMIT = 10;
	
	
	public String run() {
		// Generate all possible laminae with at most SIZE_LIMIT tiles
		int[] type = new int[SIZE_LIMIT + 1];
		for (int n = 3; (n - 1) * 4 <= SIZE_LIMIT; n++) {  // Outer square size
			for (int m = n - 2; m >= 1; m -= 2) {  // Inner square hole size
				int tiles = n * n - m * m;  // Intermediate computation may overflow, but result is correct
				if (tiles > SIZE_LIMIT)
					break;
				type[tiles]++;
			}
		}
		
		// Examine the type of each total tiling
		int count = 0;
		for (int t : type) {
			if (1 <= t && t <= TYPE_LIMIT)
				count++;
		}
		return Integer.toString(count);
	}
	
}

File: python/p174.py
# 
# Solution to Project Euler problem 174
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	SIZE_LIMIT = 1000000
	TYPE_LIMIT = 10
	
	# Generate all possible laminae with at most SIZE_LIMIT tiles
	type = [0] * (SIZE_LIMIT + 1)
	for n in range(3, SIZE_LIMIT // 4 + 2):  # Outer square size
		for m in range(n - 2, 0, -2):  # Inner square hole size
			tiles = n * n - m * m
			if tiles > SIZE_LIMIT:
				break
			type[tiles] += 1
	
	# Examine the type of each total tiling
	ans = sum(1 for t in type if 1 <= t <= TYPE_LIMIT)
	return str(ans)


if __name__ == "__main__":
	print(compute())
